### PR TITLE
Support websocket.http.response ASGI extension

### PR DESCRIPTION
**Pull request by kristjanvalur**
uvicorn now can return a response from websocket apps that choose to reject a connection with a custom response.
See https://github.com/encode/starlette/pull/2041/files for related work.

Note: The `websockets` module currently does not read the response body from a websockets upgrade requests, so we cannot effectively test that functionality, in the unittests.  The exception thrown at the client contains no body, only status code.

### CODE REVIEW

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/wsproto_impl.py:354]: Why would you send this message?

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/wsproto_impl.py:356]: I think you need to close the transport here, don't you?

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py:285]: Can't we remove this conditional and add `self.handshake_started_event.set()` on the `websocket.http.response.start`?

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py]: By the spec, `status` should always be provided.
```suggestion
```

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py:331]:
```suggestion
                    self.initial_response = (message["status"], headers, b"")
```

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py]: Access logs should be sent on `start` messages.
```suggestion
```

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py:318]:
```suggestion
                    message = cast("WebSocketResponseStartEvent", message)
                    self.logger.info(
                        '%s - "WebSocket %s" %d',
                        self.scope["client"],
                        get_path_with_query_string(self.scope),
                        message["status"],
                    )
```

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/wsproto_impl.py:276]: Can't we do the same as suggested for `websockets` i.e. remove this line from accept and close?

**Review by Kludex — DISMISSED** [review]

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/wsproto_impl.py:356]: yes, fixed.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: This is the same logic as for the `websocket.close`.  The application, when it closes from its end (with `websocket.close` or `websocket.http.response.*` also gets a close message on its input queue.

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: Ah, it is an `AttributeError` which happens if the lookup from `http.HTTPStatus` fails.  That is, if the use has provided
a code which isn't part of the enum.  This needs to be an enum, otherwise sending the response later will quietly fail.
I guess we could just allow the lookup here to fail, I was trying to be defensive.  Do you agree, or should we fallback to this code if an invalid code is provided?

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py:331]: No, the status needs to be the enum value from http.HTTPStatus.  I spent an hour debugging why sending it this way would result in an error.  Turns out that the `websockets` library is finicky:

https://github.com/aaugustin/websockets/blob/ba1ed7a65cc876ff4e0fcd4dd4711402836475e2/src/websockets/legacy/server.py#L227-L230

The attribute error raised on `status.value` and `status.phrase` is later caught here and disappears:

https://github.com/aaugustin/websockets/blob/ba1ed7a65cc876ff4e0fcd4dd4711402836475e2/src/websockets/legacy/server.py#L250-L255

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [reply uvicorn/protocols/websockets/websockets_impl.py]: > This needs to be an enum

Does `websockets` package "doesn't work" if you provide the int `message["status"]`?

Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: yes, see below.

File: tests/protocols/test_websocket.py
Comment (Kludex) [inline tests/protocols/test_websocket.py]: Unfortunately, exceptions in the `app` are ignored by the test suite, so even if you have an AssertionError, the test will never fail.

```suggestion
):
    disconnected_message = {}

    async def app(scope, receive, send):
        nonlocal disconnected_message
        assert scope["type"] == "websocket"
        assert "websocket.http.response" in scope["extensions"]

        # Pull up first recv message.
        message = await receive()
        assert message["type"] == "websocket.connect"

        # Reject the connection with a response
        response = Response(b"goodbye", status_code=400)
        await response(scope, receive, send)
        disconnected_message = await receive()

    async def websocket_session(url):
        with pytest.raises(websockets.exceptions.InvalidStatusCode) as exc_info:
            async with websockets.client.connect(url):
                pass  # pragma: no cover
        assert exc_info.value.status_code == 400
        # Websockets module currently does not read the response body from the socket.

    config = Config(
        app=app,
        ws=ws_protocol_cls,
        http=http_protocol_cls,
        lifespan="off",
        port=unused_tcp_port,
    )
    async with run_server(config):
        await websocket_session(f"ws://127.0.0.1:{unused_tcp_port}")

    assert disconnected_message == {"type": "websocket.disconnect", "code": 1006"}
```

File: tests/protocols/test_websocket.py
Comment (Kludex) [inline tests/protocols/test_websocket.py]: Same as last comment.

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py:285]: We need to somehow distinguish between the case when handshake is started with `websocket.http.response.start` as opposed to the others, because it sets us in a state where the only allowable messages are `websocket.http.response.body`. Otherwise, we cannot perform state validation.

We also cannot set the event until the body is ready, because we `await` this event (line 214) and expect any response to be ready when it is set.  This event really should be renamed, I could call it `handshake_done_event` if you agree.

I could however avoid the extra `if` clause (and indentation) and keep handling of all these handshake messages on the same level, but it would complicate the message validation logic.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: https://github.com/django/asgiref/issues/381

Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: What does `hypercorn` do? Let's follow their decision.

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py:318]: this is done.

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: done

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [reply uvicorn/protocols/websockets/websockets_impl.py]: Let's not silently replace it, tho... Let's add a log message here, please :+1:

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/wsproto_impl.py:276]: Again, we need to somehow know that we got an initial `websocket.http.response.start` message and that the only allowed messages now are `websocket.http.response.body`.  Is there some other way we can signal this?  Or do you think we shouldn't validate the messages?

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py]: Isn't it %d?
```suggestion
                        '%s - "WebSocket %s" %d',
```

Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: That was your suggested change, but yes :)

Comment (Kludex) [reply uvicorn/protocols/websockets/websockets_impl.py]: image

:sweat_smile:

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [reply uvicorn/protocols/websockets/websockets_impl.py:285]: Do you think we could benefit from another event to be used on line 214?

Comment (Kludex) [reply uvicorn/protocols/websockets/websockets_impl.py:285]: if too much work, don't worry

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: I must be getting confused. sorry.

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py:285]: I think the only reason there is an `event` there is because a task is awaiting the handshake to be performed, i.e. it needs to wait until the app has either accepted the connection or delivered a rejection reponse.  An event makes sense for that.  for other internal state logic, regular `bool` members are sufficient.  But I do think that the name of the event, at least, is not clear, so if you approve, I will rename it to `handshake_done_event`, because it indicates that a connect message has been delivered to the app, and the app has responded.

**Comment by kristjanvalur** [comment]
I don't see hypercorn sending any messages to the app:
https://github.com/pgjones/hypercorn/blob/8ae17ca68204d9718389fb3649ca0ed6ba851906/src/hypercorn/protocol/ws_stream.py#L322

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: Adding a test for this case (trying to send a message with invalid status) showed that the wsproto will cause a 500 error to be raised.  So I think it is fine if we also get that error for websockets.  I'm removing the check and adding a test to verify that the server produces a 500 error.

Comment (Kludex) [reply uvicorn/protocols/websockets/websockets_impl.py]: I'm fine with this compromise. Thanks :+1:

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: Hypercorn does not seem to send anything to the app when it closes the connection.
https://github.com/pgjones/hypercorn/blob/8ae17ca68204d9718389fb3649ca0ed6ba851906/src/hypercorn/protocol/ws_stream.py#L322

Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: Let's wait for Andrew's reply on the issue above, but to me, you shouldn't expect a message when sending a response. :thinking:

**Comment by Kludex** [comment]
I don't have more comments here. Beautiful PR, beautiful tests. :+1:

Thanks @kristjanvalur :pray: 

For me, before approving:
- [ ] Check if all lines are covered.
- [ ] Check Andrew's reply to the disconnect message.

**Comment by kristjanvalur** [comment]
Please see test `test_server_reject_connection_with_invalid_msg()`

I am unable to silence the wsproto internal error which happens when it is trying to send a 500 error.

1. wsproto will start a response as soon as it gets the `.start` message
2. It then gets an invalid message and attempts to send a 500 response
3. but it cannot because it has already started a response.
4. An exception is not raised via the async `send()` method but somewhere else which I cannot fully understand, so the test fails.

I've added a skip for the WSProto case, but maybe there is a better solution?
Skipping is not ideal, because this test was designed to increase coverage.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: Yes, I kind of agree.  In case that turns out to be the case, should we also remove the message fore the `websocket.close` case?

Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: I thought I had asked that here: https://github.com/django/asgiref/issues/364

But it's actually a different case... :thinking: There was also a discussion here: https://github.com/django/asgiref/issues/340.

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py]: Eh... :thinking:

Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: I see!  I didn't notice this one.  Let me take another look at this.  Silly me.

Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: Undone.  sorry for the noise.

**Comment by kristjanvalur** [comment]
A separate question:  The `websockets` library does not read the response body from a websocket upgrade request.  Do you think it is worth making a PR against that repo to add that functionality?

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py:354]: Andrew replied. It's good to send the disconnect message. :+1:

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (Kludex) [inline uvicorn/protocols/websockets/websockets_impl.py]: This doesn't make any difference, does it? After the response is sent, the connection is closed.
```suggestion
```

**Comment by Kludex** [comment]
> A separate question: The `websockets` library does not read the response body from a websocket upgrade request. Do you think it is worth making a PR against that repo to add that functionality?

That would be great :+1: 

First ask the maintainer if he would accept such a PR, please :eyes:

**Comment by Kludex** [comment]
> Please see test `test_server_reject_connection_with_invalid_msg()`
> 
> I am unable to silence the wsproto internal error which happens when it is trying to send a 500 error.
> 
> 1. wsproto will start a response as soon as it gets the `.start` message
> 2. It then gets an invalid message and attempts to send a 500 response
> 3. but it cannot because it has already started a response.
> 4. An exception is not raised via the async `send()` method but somewhere else which I cannot fully understand, so the test fails.
> 
> I've added a skip for the WSProto case, but maybe there is a better solution? Skipping is not ideal, because this test was designed to increase coverage.

We should have the analogous behavior that we have when using `http.response.start` and `http.response.body`. What's the behavior with those events (I think they should be the same on httptools and h11)?

File: uvicorn/protocols/websockets/websockets_impl.py
Comment (kristjanvalur) [reply uvicorn/protocols/websockets/websockets_impl.py]: No, it's just me not wanting to leave already processed data lying around.  I'll remove it.

**Comment by kristjanvalur** [comment]
> We should have the analogous behavior that we have when using `http.response.start` and `http.response.body`. What's the behavior with those events (I think they should be the same on httptools and h11)?

Okay, I looked at the similar test for http (`test_duplicate_start_message`) and it, in fact, asserts that it **does not** get a 500 response, precisely because a 200 response has been previously written into the buffer.

for websockets, and the `websockets` protocol, we can't send the response incrementally.  Instead, we prepare it and return it all at once.  That means that we can in fact issue a 500 error in this case.

It is not a problem to modify the test for both cases.
The problem remains for me, how to run the test with wsproto, because it fails in some weird way.  I'll try some things.

**Comment by kristjanvalur** [comment]
> It is not a problem to modify the test for both cases.
> The problem remains for me, how to run the test with wsproto, because it fails in some weird way. I'll try some things.

Turns out that we needed extra tests to check if response had been started before attempting to send a 500 response, similar to what is done in h11.

**Comment by kristjanvalur** [comment]
seems like workflows are disabled 🤔

**Comment by Kludex** [comment]
I cancelled a pipeline yesterday that it was not able to stop, maybe it's related to that. I don't know how to enable it again here.

**Comment by kristjanvalur** [comment]
Weird, yes, the last time a workflow ran in this PR it timed out.

**Comment by Kludex** [comment]
You can try to reopen a new PR. 🤔

**Pull request closed without merging**
